source/common/ssl/ssl_socket.cc

```diff
@@ -402,39 +402,67 @@ std::string SslSocket::subjectLocalCertificate() const {
   return getSubjectFromCertificate(cert);
 }
 
+namespace {
+SslSocketFactoryStats generateStats(const std::string& prefix, Stats::Scope& store) {
+  return {
+      ALL_SSL_SOCKET_FACTORY_STATS(POOL_COUNTER_PREFIX(store, prefix + "_ssl_socket_factory."))};
+}
+} // namespace
+
 ClientSslSocketFactory::ClientSslSocketFactory(ClientContextConfigPtr config,
                                                Ssl::ContextManager& manager,
                                                Stats::Scope& stats_scope)
-    : manager_(manager), stats_scope_(stats_scope), config_(std::move(config)),
-      ssl_ctx_(manager_.createSslClientContext(stats_scope_, *config_)) {}
+    : manager_(manager), stats_scope_(stats_scope), stats_(generateStats("client", stats_scope)),
+      config_(std::move(config)),
+      ssl_ctx_(manager_.createSslClientContext(stats_scope_, *config_)) {
+  config_->setSecretUpdateCallback([this]() { onAddOrUpdateSecret(); });
+}
 
 Network::TransportSocketPtr ClientSslSocketFactory::createTransportSocket() const {
   if (ssl_ctx_) {
```

Comment:
while locking is not neccessary, you have to capture ssl_ctx_ into a local variable here, `onAddOrUpdateSecret` may overwrite it since it's on main thread.

Reply:
Good point. Thanks! Fixed this issue and added comment above.